README
sqlite3 (Sqlite3) module for Qore
version 1.0 by Petr Vanek, David Nichols

Requires qore 0.7.0+ and sqlite3 headers and library to build

The driver supports the following features:
* all sqlite3 afinities (datatypes) are supported, except "date" types.
* "date" types are supported only for writing from Qore. See Sqlite3
  docs for more infromations about its dates handling.
* multi-dimensional arrays are not supported for binding and retrieving
* is thread-safe
* transaction management is supported
* transparent character encoding conversion is supported if necessary

The driver's name is "sqlite3".
ex: $db = new Datasource("sqlite3");

The Qore sqlite3 driver uses binary communication with Sqlite3 files

Documentation in docs/sqlite3-module-doc.html

test/example scripts are stored in test/ subdirectory


BUILD AND INSTALL
-----------------

Sqlite3 DBI module requires Sqlite3 library and headers.
If your Sqlite3 library and includes are in a non-standard location you can
use the --with-sqlite3 configure option or set the SQLITE3_DIR environment
variable.

configure
make
sudo make install

The configure script will find out where your qore module directory is found
and set this for the install directory.

BUILD FROM SVN
--------------

reconf.sh
configure
make
sudo make install


ChangeLog
2010-06-18 12:29  david_nichols

	* qore-sqlite3-module.spec: updated spec file to assume sqlite3*
	  package name if not redhat (now works on mandriva)

2010-06-17 15:25  david_nichols

	* qore-sqlite3-module.spec: minor fixes in spec files (typos +
	  rpmlint fixes)

2010-06-15 18:49  david_nichols

	* docs/sqlite3-module-doc.html: updated tikiwiki and added updated
	  color scheme

2010-04-27 17:55  subik

	* src/sqlite3executor.cc, src/sqlite3executor.h,
	  test/sqlite3test-basic.q, test/sqlite3test-threading.q: test are
	  silent now; exceptions are more descriptive

2010-04-16 09:32  subik

	* RELEASE-NOTES, configure.ac, qore-sqlite3-module.spec,
	  src/sqlite3module.cc: sql drivers versions bump to spot execRaw.
	  execRaw feature promoted to all drivers' feature lists. execRaw
	  constant naming unified.

2010-02-22 09:31  subik

	* src/sqlite3executor.cc, src/sqlite3executor.h,
	  src/sqlite3module.cc: backward compatibility for execRaw

2010-02-14 07:39  subik

	* src/sqlite3executor.cc, src/sqlite3executor.h,
	  src/sqlite3module.cc, test/sqlite3test-basic.q: implementation of
	  execRaw

2009-12-09 12:58  david_nichols

	* configure.ac: fixed all modules to use decimal comparisons with
	  the qore version string instead of octal so that they will still
	  compile with qore v0.8.0 and up...

2009-11-22 06:03  david_nichols

	* configure.ac: updated configure to assume 64-bit build if
	  host_cpu = "amd64"

2009-10-30 08:22  david_nichols

	* configure.ac: updated all module configure scripts to use gnutar
	  on darwin to avoid picking up OSX resource files in source
	  releases

2009-10-05 10:36  david_nichols

	* ChangeLog: updated changelog

2009-09-21 11:07  david_nichols

	* configure.ac: updated module configure scripts to select 64-bit
	  target mode by default on snow leopard i386

2009-09-10 06:49  david_nichols

	* configure.ac: updated configure.ac to properly compile in 32-bit
	  mode on OSX snow leopard (10.6)

2009-07-20 13:16  david_nichols

	* configure.ac, qore-sqlite3-module.spec: added opensuse 11.2
	  detection to spec file, fixed rhel detection in spec files,
	  updated configure.ac to work with older versions of autotools

2009-07-03 15:02  subik

	* test/sqlite3test-basic.q: make test less talkative without
	  verbose option

2009-07-03 08:07  subik

	* ., AUTHORS, COPYING, ChangeLog, INSTALL, Makefile.am, NEWS,
	  README, RELEASE-NOTES, configure.ac, docs, m4,
	  qore-sqlite3-module.spec, reconf.sh, src, test: standard dir
	  structure



src/sqlite3module.h
/*
  sqlite3module.cc

  Qore Programming Language

  Copyright 2003 - 2009 Qore Technologies, s.r.o <http://qore.org>

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef SQLITE3MODULE_H
#define SQLITE3MODULE_H

QoreStringNode *qore_sqlite3_module_init();
void qore_sqlite3_module_ns_init(QoreNamespace *rns, QoreNamespace *qns);
void qore_sqlite3_module_delete(void);

#endif


src/sqlite3executor.h
/*
  sqlite3executor.h

  Qore Programming Language

  Copyright 2003 - 2009 Qore Technologies, s.r.o <http://qore.org>

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef SQLITE3EXECUTOR_H
#define SQLITE3EXECUTOR_H

#include <sqlite3.h>
#include <qore/Qore.h>


/*! \brief Data operations class.
It will be created for every request from sqlite3module functions
to keep statement related data clean and separated from QoreSqlite3Connection.
\author Petr Vanek <petr.vanek@qoretechnologies.com>
*/
class QoreSqlite3Executor
{
    public:
        QoreSqlite3Executor(sqlite3 * handler, ExceptionSink *xsink);
        ~QoreSqlite3Executor();

        /*! \brief Implementation for Qore DB API exec().
        It's primarily used for DDL/INSERT/UPDATE/DELETE statemets, but
        it can handle all stuff as it's calling select() method.
        \param ds a Datasource reference from Qore API.
        \param qstr a SQL statement from Qore API.
        \param args a list with bindable parameters from Qore API.
        \param xsink exception handler.
        \retval AbstractQoreNode* hash for SELECTs, count of the affected
                rows for CRUD stuff.
        */
        AbstractQoreNode * exec(Datasource *ds,
                                const QoreString *qstr,
                                const QoreListNode *args,
                                ExceptionSink *xsink);

#ifdef _QORE_HAS_DBI_EXECRAW
        /*! \brief Implementation for Qore DB API execRaw().
        It's primarily used for DDL statemets, but
        it can handle all stuff. But there is no variable binding.
        \param ds a Datasource reference from Qore API.
        \param qstr a SQL statement from Qore API.
        \param xsink exception handler.
        \retval AbstractQoreNode* hash for SELECTs, count of the affected
                rows for CRUD stuff.
        */
        AbstractQoreNode * execRaw(Datasource *ds,
                                const QoreString *qstr,
                                ExceptionSink *xsink);
#endif

        /*! \brief Implementation for Qore DB API select().
        \param ds a Datasource reference from Qore API.
        \param qstr a SQL statement from Qore API.
        \param args a list with bindable parameters from Qore API.
        \param xsink exception handler.
        \retval AbstractQoreNode* QoreHashNode with coluns as keys, data as lists.
        */
        AbstractQoreNode * select(
            Datasource *ds,
            const QoreString *qstr,
            const QoreListNode *args,
            ExceptionSink *xsink);

        /*! \brief Implementation for Qore DB API selectRows().
        \param ds a Datasource reference from Qore API.
        \param qstr a SQL statement from Qore API.
        \param args a list with bindable parameters from Qore API.
        \param xsink exception handler.
        \retval AbstractQoreNode* QoreListNode with the list of hashes.
        */
        AbstractQoreNode * select_rows(
            Datasource *ds,
            const QoreString *qstr,
            const QoreListNode *args,
            ExceptionSink *xsink);

    private:
        //! Everywhere used current sqlite3 connection.
        sqlite3 * m_handler;
        //! Really used parameters list for statement (%s etc. filtered out)
        ReferenceHolder<QoreListNode> m_realArgs;

        /*! \brief Universal Sqlite3 to Qore nodes conversion.
        \param stmt a reference for sqlite3 SQL statement. It has to be
                    prepared and fetched already.
        \param index index of column to retrieve.
        \retval AbstractQoreNode* Qore node type by its sqlite afinity.
                    Possible nodes: QoreBigIntNode, QoreFloatNode, BinaryNode,
                                    QoreStringNode, and null().
        */
        AbstractQoreNode * columnValue(sqlite3_stmt * stmt, int index);

        /*! \brief Prepare a string (SQL statement) for variables binding.
        It performs some Qore DB bind syntax validity checks too.
        \param str a mutable Qore string with SQL statement
        \param args a list with bindable parameters
        \param xsink exceptions handler
        \retval bool true on success, false on error
        */
        bool parseForBind(QoreString & str, const QoreListNode *args, ExceptionSink *xsink);

        /*! \brief Bind arguments to the SQL statement.
        \param stmt a prepared sqlite3 statement.
        \param args a list with bindable parameters
        \param xsink exception handler
        \retval bool true on success, false on error
        */
        bool bindParameters(sqlite3_stmt * stmt, ExceptionSink *xsink);

        /*! \brief Internal implementation of select() DB API.
        \param ds a Datasource reference from Qore API.
        \param qstr a SQL statement from Qore API.
        \param args a list with bindable parameters from Qore API.
        \param binding flag if it should allow variable binding (true) or not (false)
        \param xsink exception handler.
        \retval AbstractQoreNode* QoreHashNode with coluns as keys, data as lists.
        */
        AbstractQoreNode * select_internal(
            Datasource *ds,
            const QoreString *qstr,
            const QoreListNode *args,
            bool binding,
            const char * calltype,
            ExceptionSink *xsink);
};

#endif


src/sqlite3executor.cc
/*
  sqlite3executor.cc

  Qore Programming Language

  Copyright 2003 - 2009 Qore Technologies, s.r.o <http://qore.org>

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "sqlite3executor.h"


QoreSqlite3Executor::QoreSqlite3Executor(sqlite3 * handle,
                                          ExceptionSink *xsink)
        : m_handler(handle),
        m_realArgs(new QoreListNode(), xsink)
{
}

QoreSqlite3Executor::~QoreSqlite3Executor()
{
}

AbstractQoreNode * QoreSqlite3Executor::columnValue(sqlite3_stmt * stmt, int index)
{
    int columnType = sqlite3_column_type(stmt, index);

    switch (columnType)
    {
    case SQLITE_INTEGER:
        return new QoreBigIntNode(sqlite3_column_int(stmt, index));
        break;
    case SQLITE_FLOAT:
        return new QoreFloatNode(sqlite3_column_double(stmt, index));
        break;
    case SQLITE_BLOB:
    {
        int nBlob = sqlite3_column_bytes(stmt, index);
        unsigned char *zBlob = (unsigned char *)malloc(nBlob);
        memcpy(zBlob, sqlite3_column_blob(stmt, index), nBlob);
        BinaryNode * bn = new BinaryNode(zBlob, nBlob);
        return bn;
        break;
    }
    case SQLITE_NULL:
        return null();
        break;
    case SQLITE_TEXT:
    default:
        return new QoreStringNode((const char*)sqlite3_column_text(stmt, index));
        break;
    };
    return 0;
}

AbstractQoreNode * QoreSqlite3Executor::exec(
    Datasource *ds,
    const QoreString *qstr,
    const QoreListNode *args,
    ExceptionSink *xsink)
{
    ReferenceHolder<QoreHashNode> hash(reinterpret_cast<QoreHashNode*>(select_internal(ds, qstr, args, true, "DBI:SQLITE3:EXEC", xsink)), xsink);
    if (*xsink)
        return 0;

    if (hash->size() > 0)
        return hash.release();

    return new QoreBigIntNode(sqlite3_changes(m_handler));
}

#ifdef _QORE_HAS_DBI_EXECRAW
AbstractQoreNode * QoreSqlite3Executor::execRaw(
    Datasource *ds,
    const QoreString *qstr,
    ExceptionSink *xsink)
{
    ReferenceHolder<QoreHashNode> hash(reinterpret_cast<QoreHashNode*>(select_internal(ds, qstr, 0, false, "DBI:SQLITE3:EXECRAW", xsink)), xsink);
    if (*xsink)
        return 0;

    if (hash->size() > 0)
        return hash.release();

    return new QoreBigIntNode(sqlite3_changes(m_handler));
}
#endif
            
AbstractQoreNode * QoreSqlite3Executor::select(
    Datasource *ds,
    const QoreString *qstr,
    const QoreListNode *args,
    ExceptionSink *xsink)
{
    ReferenceHolder<QoreHashNode> hash(reinterpret_cast<QoreHashNode*>(select_internal(ds, qstr, args, true, "DBI:SQLITE3:SELECT", xsink)), xsink);
    if (*xsink)
        return 0;

    if (hash->size() > 0)
        return hash.release();

    return new QoreBigIntNode(sqlite3_changes(m_handler));
}

AbstractQoreNode * QoreSqlite3Executor::select_rows(
    Datasource *ds,
    const QoreString *qstr,
    const QoreListNode *args,
    ExceptionSink *xsink)
{
    QoreString statement(qstr);
    if (!parseForBind(statement, args, xsink))
        return xsink->raiseException("DBI:SQLITE3:SELECT_ROWS",
                                      "failed to parse bind variables");

    sqlite3_stmt *stmt;
    int rc = sqlite3_prepare_v2(m_handler, statement.getBuffer(), -1, &stmt, 0);
    if (rc != SQLITE_OK)
        return xsink->raiseException("DBI:SQLITE3:SELECT_ROWS",
                                      "sqlite3 error: %s", sqlite3_errmsg(m_handler));

    if (!bindParameters(stmt, xsink))
        return xsink->raiseException("DBI:SQLITE3:SELECT_ROWS",
                                      "failed to bind variables");

    ReferenceHolder<QoreListNode> res(new QoreListNode(), xsink);

    while (sqlite3_step(stmt) == SQLITE_ROW)
    {
        QoreHashNode * head = new QoreHashNode();

        for (int i = 0; i < sqlite3_column_count(stmt); ++i)
        {
            head->setKeyValue(sqlite3_column_name(stmt, i),
                              columnValue(stmt, i),
                              xsink);
        }
        res->push(head);
    }

    sqlite3_finalize(stmt);
    return res.release();
}

bool QoreSqlite3Executor::parseForBind(QoreString & str,
                                        const QoreListNode *args,
                                        ExceptionSink *xsink)
{
    char quote = 0;
    const char *p = str.getBuffer();
    QoreString tmp;
    int index = 0;
    int nParams = 0;
    int nIterations = 0;
    int argsSize = (args) ? args->size() : 0;

//     m_realArgs = new QoreListNode();

    while (*p)
    {
        if (!quote && (*p) != '%')
        {
            ++p;
            continue;
        }
        // found value marker

        int offset = p - str.getBuffer();

        p++;
        const AbstractQoreNode *v = args ? args->retrieve_entry(index++) : NULL;
        ++nIterations;
        if (v == 0)
        {
            xsink->raiseException("DBI-EXEC-PARSE-EXCEPTION",
                              "Parameter count mismatch (args: %d ; processed: %d)", argsSize, nIterations);
            return false;
        }

        if ((*p) == 'd')
        {
            DBI_concat_numeric(&tmp, v);
            str.replace(offset, 2, &tmp);
            p = str.getBuffer() + offset + tmp.strlen();
            tmp.clear();
            continue;
        }
        if ((*p) == 's')
        {
            if (DBI_concat_string(&tmp, v, xsink))
                return false;
            str.replace(offset, 2, &tmp);
            p = str.getBuffer() + offset + tmp.strlen();
            tmp.clear();
            continue;
        }
        if ((*p) != 'v')
        {
            xsink->raiseException("DBI-EXEC-PARSE-EXCEPTION",
                                  "invalid value specification (expecting '%v' or '%%d', got %%%c)", *p);
            return false;
        }
        ++p;
        if (isalpha(*p))
        {
            xsink->raiseException("DBI-EXEC-PARSE-EXCEPTION",
                                  "invalid value specification (expecting '%v' or '%%d', got %%v%c*)", *p);
            return false;
        }

        // replace value marker with "?<num>" - sqlite3 binding
        // find byte offset in case string buffer is reallocated with replace()
        ++nParams;
        tmp.sprintf("?%d", nParams);
        str.replace(offset, 2, &tmp);
        p = str.getBuffer() + offset + tmp.strlen();
        tmp.clear();

        m_realArgs->push(v->realCopy());

        if (((*p) == '\'') || ((*p) == '\"'))
        {
            if (!quote)
                quote = *p;
            else if (v)
                quote = 0;
            ++p;
        }
    }

    if (nIterations != argsSize)
    {
        xsink->raiseException("DBI-EXEC-PARSE-EXCEPTION",
                              "Parameter count mismatch (args: %d ; processed: %d)", argsSize, nIterations);
        return false;
    }

    return true;
}

bool QoreSqlite3Executor::bindParameters(sqlite3_stmt * stmt, ExceptionSink *xsink)
{
    qore_type_t argType;
    const AbstractQoreNode * arg;

    for (int i = 0; i < sqlite3_bind_parameter_count(stmt); ++i)
    {
        arg = m_realArgs->retrieve_entry(i);

        if (!arg)
        {
            xsink->raiseException("DBI-EXEC-BIND-EXCEPTION",
                                   "unknown/missing argument to bind.");
            return false;
        }

        argType = arg->getType();

        switch (argType)
        {
        case NT_INT:
            if (SQLITE_OK != sqlite3_bind_int64(stmt, i+1, reinterpret_cast<const QoreBigIntNode*>(arg)->val))
            {
                xsink->raiseException("DBI-EXEC-BIND-EXCEPTION",
                                       "Cannot bind integer.");
                return false;
            }
            break;
        case NT_FLOAT:
            if (SQLITE_OK != sqlite3_bind_double(stmt, i+1, reinterpret_cast<const QoreFloatNode*>(arg)->f))
            {
                xsink->raiseException("DBI-EXEC-BIND-EXCEPTION",
                                       "Cannot bind double/fload.");
                return false;
            }
            break;
        case NT_STRING:
        {
            bool del;
            QoreString s(reinterpret_cast<const QoreStringNode*>(arg)->getStringRepresentation(del));
            if (SQLITE_OK != sqlite3_bind_text(stmt, i+1, s.getBuffer(), s.strlen(), SQLITE_TRANSIENT))
            {
                xsink->raiseException("DBI-EXEC-BIND-EXCEPTION",
                                       "Cannot bind string.");
                return false;
            }
            break;
        }
        case NT_BOOLEAN:
            if (SQLITE_OK != sqlite3_bind_double(stmt, i+1, reinterpret_cast<const QoreBoolNode*>(arg)->getValue()))
            {
                xsink->raiseException("DBI-EXEC-BIND-EXCEPTION",
                                       "Cannot bind double/fload.");
                return false;
            }
            break;
//         case NT_DATE:
//             // TODO/FIXME: date
//             xsink->raiseException("DBI-EXEC-BIND-EXCEPTION", "Date Time handling is not implemented yet");
//             return false;
        case NT_BINARY:
        {
            const BinaryNode * b = reinterpret_cast<const BinaryNode*>(arg);
            if (SQLITE_OK != sqlite3_bind_blob(stmt, i+1, b->getPtr(), b->size(), NULL))
            {
                xsink->raiseException("DBI-EXEC-BIND-EXCEPTION",
                                       "Cannot bind BLOB.");
                return false;
            }
            break;
        }
        default:
            xsink->raiseException("DBI-EXEC-BIND-EXCEPTION",
                                  "Cannot bind '%s' for Sqlite3 by its design.", arg->getTypeName());
            return false;
        } // switch
    } // for
    return true;
}

AbstractQoreNode * QoreSqlite3Executor::select_internal(
            Datasource *ds,
            const QoreString *qstr,
            const QoreListNode *args,
            bool binding,
            const char * calltype,
            ExceptionSink *xsink)
{
    QoreString statement(qstr);//new QoreString(qstr);
    if (binding) {
        if (!parseForBind(statement, args, xsink))
            return xsink->raiseException(calltype,
                                         "failed to parse bind variables");
    }

    sqlite3_stmt *stmt;

    int rc = sqlite3_prepare_v2(m_handler, statement.getBuffer(), -1, &stmt, 0);
    if (rc != SQLITE_OK)
        return xsink->raiseException(calltype,
                                      "sqlite3 error: %s", sqlite3_errmsg(m_handler));

    if (binding) {
        if (!bindParameters(stmt, xsink))
            return xsink->raiseException(calltype,
                                         "failed to bind variables");
    }

    // columns as keys
    ReferenceHolder<QoreHashNode> hash(new QoreHashNode(), xsink);

    for (int i = 0; i < sqlite3_column_count(stmt); ++i)
        hash->setKeyValue(sqlite3_column_name(stmt, i), new QoreListNode(), xsink);

    // fetch the results
    while (sqlite3_step(stmt) == SQLITE_ROW)
    {
        for (int i = 0; i < sqlite3_column_count(stmt); ++i)
        {
            QoreListNode * node = reinterpret_cast<QoreListNode *>(hash->getKeyValue(sqlite3_column_name(stmt, i)));
            node->push(columnValue(stmt, i));
        }
    }

    sqlite3_finalize(stmt);
    return hash.release();
}



src/sqlite3connection.h
/*
  sqlite3connection.h

  Qore Programming Language

  Copyright 2003 - 2009 Qore Technologies, s.r.o <http://qore.org>

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef SQLITE3CONNECTION_H
#define SQLITE3CONNECTION_H

#include <sqlite3.h>
#include <qore/Qore.h>


/*! \brief A Qore ready wrapper for Sqlite3 API.
There is only one instance of this class in this module.
All select/exec depending stuff is located in QoreSqlite3Executor,
the class with created instances on demand for each statement.
\author Petr Vanek <petr.vanek@qoretechnologies.com>
*/
class QoreSqlite3Connection
{
public:
    /*! \brief Cosntruct a Sqlite3 driver for Qore module.
    \param sqlite3 * a reference to already inicialized sqlite3 handler.
    */
    QoreSqlite3Connection(sqlite3 * handler);

    ~QoreSqlite3Connection() {};

    /*! \brief Public access to the DB conection handler.
    \retval sqlite3* sqlite3 handler.
    */
    sqlite3 * handler() { return m_handler; };

    /*! \brief Start a transaction.
    Error checking is left for sqlite3 engine.
    \retval bool true for success; false for any error.
    */
    bool begin(ExceptionSink *xsink);

    /*! \brief Preforms a COMMIT statement without any transaction checking.
    Error checking is left for sqlite3 engine.
    \retval bool true for success; false for any error.
    */
    bool commit(ExceptionSink *xsink);

    /*! \brief Preforms a ROLLBACK statement without any transaction checking.
    Error checking is left for sqlite3 engine.
    \retval bool true for success; false for any error.
    */
    bool rollback(ExceptionSink *xsink);

    /*! \brief Close active sqlite3 handler.
    \retval bool true on success; false on error.
    */
    bool close();

    /*! \brief Get sqlite3 backend version.
    Sqlite DB engine is "serverless" so there is only this method
    as we don't need any "client version". Qore DB API is using this
    method for both get*Version() methods.
    \retval char* A string with Sqlite3 library in use (format: X.Y.Z)
    */
    char * getServerVersion();

private:
    //! Everywhere used current sqlite3 connection.
    sqlite3 * m_handler;
};

#endif


src/sqlite3module.cc
/*
  sqlite3module.cc

  Qore Programming Language

  Copyright 2003 - 2009 Qore Technologies, s.r.o <http://qore.org>

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <strings.h>

#include <qore/Qore.h>

#include "sqlite3module.h"
#include "sqlite3connection.h"
#include "sqlite3executor.h"
#include "config.h"


#ifndef QORE_MONOLITHIC
DLLEXPORT char qore_module_name[] = "sqlite3";
DLLEXPORT char qore_module_version[] = PACKAGE_VERSION;
DLLEXPORT char qore_module_description[] = "Sqlite3 database driver";
DLLEXPORT char qore_module_author[] = "Petr Vanek";
DLLEXPORT char qore_module_url[] = "http://www.qore.org";
DLLEXPORT int qore_module_api_major = QORE_MODULE_API_MAJOR;
DLLEXPORT int qore_module_api_minor = QORE_MODULE_API_MINOR;
DLLEXPORT qore_module_init_t qore_module_init = qore_sqlite3_module_init;
DLLEXPORT qore_module_ns_init_t qore_module_ns_init = qore_sqlite3_module_ns_init;
DLLEXPORT qore_module_delete_t qore_module_delete = qore_sqlite3_module_delete;

DLLEXPORT qore_license_t qore_module_license = QL_LGPL;
#endif


DBIDriver *DBID_SQLITE3 = 0;

// this is the thread key that will tell us
// if the current thread has been initialized
// for sqlite threading
static pthread_key_t ptk_sqlite3;


static inline void checkInit()
{
    if (!pthread_getspecific(ptk_sqlite3))
    {
        pthread_setspecific(ptk_sqlite3, (void *)1);
    }
}

static void sqlite3_thread_cleanup(void *unused)
{
//     if (pthread_getspecific(ptk_sqlite3))
}



static sqlite3 *qore_sqlite3_init(Datasource *ds, ExceptionSink *xsink)
{
    if (!ds->getDBName())
    {
        xsink->raiseException("DATASOURCE-MISSING-DBNAME",
                              "Datasource has an empty dbname parameter");
        return 0;
    }

    // TODO/FIXME: better encoding handling (but sqlite is utf8 mainly)
    ds->setDBEncoding("utf8");
    ds->setQoreEncoding("utf8");

    sqlite3 *db;
    int ret = sqlite3_open(ds->getDBName(), &db);
    if (ret != SQLITE_OK)
    {
        xsink->raiseException("DBI:SQLITE3:CONNECT-ERROR",
                              "cannot open %s", ds->getDBName());
        return 0;
    }

    if (!db)
    {
        xsink->outOfMemory();
        return 0;
    }

    return db;
}

static int qore_sqlite3_commit(Datasource *ds, ExceptionSink *xsink)
{
    checkInit();
    QoreSqlite3Connection *d =(QoreSqlite3Connection *)ds->getPrivateData();
    return d->commit(xsink) ? 0 : -1;
}

static int qore_sqlite3_rollback(Datasource *ds, ExceptionSink *xsink)
{
    checkInit();
    QoreSqlite3Connection *d =(QoreSqlite3Connection *)ds->getPrivateData();
    return d->rollback(xsink) ? 0 : -1;
}

static AbstractQoreNode *qore_sqlite3_select_rows(Datasource *ds,
                                                   const QoreString *qstr,
                                                   const QoreListNode *args,
                                                   ExceptionSink *xsink)
{
    checkInit();
    QoreSqlite3Connection *d = (QoreSqlite3Connection *)ds->getPrivateData();
    QoreSqlite3Executor exec(d->handler(), xsink);
    return exec.select_rows(ds, qstr, args, xsink);
}

static AbstractQoreNode *qore_sqlite3_select(Datasource *ds,
                                              const QoreString *qstr,
                                              const QoreListNode *args,
                                              ExceptionSink *xsink)
{
    checkInit();
    QoreSqlite3Connection *d = (QoreSqlite3Connection *)ds->getPrivateData();
    QoreSqlite3Executor exec(d->handler(), xsink);
    return exec.select(ds, qstr, args, xsink);
}

static AbstractQoreNode *qore_sqlite3_exec(Datasource *ds,
                                            const QoreString *qstr,
                                            const QoreListNode *args,
                                            ExceptionSink *xsink)
{
    checkInit();
    QoreSqlite3Connection *d = (QoreSqlite3Connection *)ds->getPrivateData();
    QoreSqlite3Executor exec(d->handler(), xsink);
    return exec.exec(ds, qstr, args, xsink);
}

#ifdef _QORE_HAS_DBI_EXECRAW
static AbstractQoreNode *qore_sqlite3_exec_raw(Datasource *ds,
                                            const QoreString *qstr,
                                            ExceptionSink *xsink)
{
    checkInit();
    QoreSqlite3Connection *d = (QoreSqlite3Connection *)ds->getPrivateData();
    QoreSqlite3Executor exec(d->handler(), xsink);
    return exec.execRaw(ds, qstr, xsink);
}
#endif

static int qore_sqlite3_open_datasource(Datasource *ds, ExceptionSink *xsink)
{
// 	printf("open datasource\n");
    checkInit();

    sqlite3 *db = qore_sqlite3_init(ds, xsink);
    if (!db)
        return -1;

    QoreSqlite3Connection *d_sqlite3 = new QoreSqlite3Connection(db);
    ds->setPrivateData((void *)d_sqlite3);

    return 0;
}

static int qore_sqlite3_close_datasource(Datasource *ds)
{
    QORE_TRACE("qore_sqlite3_close_datasource()");

    checkInit();
    QoreSqlite3Connection *d = (QoreSqlite3Connection *)ds->getPrivateData();

    if (!d->close())
        return -1;

    delete d;
    ds->setPrivateData(NULL);

    return 0;
}

static AbstractQoreNode *qore_sqlite3_get_server_version(Datasource *ds,
                                                          ExceptionSink *xsink)
{
    checkInit();
    QoreSqlite3Connection *d = (QoreSqlite3Connection *)ds->getPrivateData();
    return new QoreStringNode(d->getServerVersion());
}

static AbstractQoreNode *qore_sqlite3_get_client_version(const Datasource *ds,
                                                          ExceptionSink *xsink)
{
    checkInit();
    QoreSqlite3Connection *d = (QoreSqlite3Connection *)ds->getPrivateData();
    return new QoreStringNode(d->getServerVersion());
}

static int qore_sqlite3_begin_tranaction(Datasource *ds,
                                          ExceptionSink *xsink)
{
    checkInit();
    QoreSqlite3Connection *d =(QoreSqlite3Connection *)ds->getPrivateData();
    return d->begin(xsink) ? 0 : -1;
}

static int qore_sqlite3_abort_transaction_start(Datasource *ds,
                                                 ExceptionSink *xsink)
{
    return qore_sqlite3_rollback(ds, xsink);
}

#ifdef _QORE_HAS_DBI_EXECRAW
#define SQL3_DBI_CAP_HAS_EXECRAW DBI_CAP_HAS_EXECRAW
#else
#define SQL3_DBI_CAP_HAS_EXECRAW 0
#endif

QoreStringNode *qore_sqlite3_module_init()
{
    pthread_key_create(&ptk_sqlite3, NULL);
    tclist.push(sqlite3_thread_cleanup, NULL);

    // populate the method list structure with the method pointers
    qore_dbi_method_list methods;
    methods.add(QDBI_METHOD_OPEN,               qore_sqlite3_open_datasource);
    methods.add(QDBI_METHOD_CLOSE,              qore_sqlite3_close_datasource);
    methods.add(QDBI_METHOD_SELECT,             qore_sqlite3_select);
    methods.add(QDBI_METHOD_SELECT_ROWS,        qore_sqlite3_select_rows);
    methods.add(QDBI_METHOD_EXEC,               qore_sqlite3_exec);
#ifdef _QORE_HAS_DBI_EXECRAW
    methods.add(QDBI_METHOD_EXECRAW,            qore_sqlite3_exec_raw);
#endif
    methods.add(QDBI_METHOD_COMMIT,             qore_sqlite3_commit);
    methods.add(QDBI_METHOD_ROLLBACK,           qore_sqlite3_rollback);
    methods.add(QDBI_METHOD_GET_SERVER_VERSION, qore_sqlite3_get_server_version);
    methods.add(QDBI_METHOD_GET_CLIENT_VERSION, qore_sqlite3_get_client_version);
    methods.add(QDBI_METHOD_BEGIN_TRANSACTION,  qore_sqlite3_begin_tranaction);
    methods.add(QDBI_METHOD_ABORT_TRANSACTION_START, qore_sqlite3_abort_transaction_start);

    // register database functions with DBI subsystem
    DBID_SQLITE3 = DBI.registerDriver("sqlite3", methods,
                                       DBI_CAP_LOB_SUPPORT | DBI_CAP_TRANSACTION_MANAGEMENT
                                       | DBI_CAP_BIND_BY_PLACEHOLDER | DBI_CAP_BIND_BY_VALUE
                                       | SQL3_DBI_CAP_HAS_EXECRAW
                                     );

    return 0;
}

void qore_sqlite3_module_ns_init(QoreNamespace *rns, QoreNamespace *qns)
{
    QORE_TRACE("qore_sqlite3_module_ns_init()");
    // nothing to do at the moment
}

void qore_sqlite3_module_delete()
{
    QORE_TRACE("qore_sqlite3_module_delete()");

    // cleanup any thread data
    tclist.pop(1);
    // delete thread key
    pthread_key_delete(ptk_sqlite3);
}
